code/modules/events/apc_short.dm

```diff
@@ -27,51 +27,105 @@
+			var/list/skipped_areas_apc = list(
+			/area/engine/engineering,
+			/area/turret_protected/ai)
+			if(announce)
+				GLOB.event_announcement.Announce("Power failure detected in [station_name()]'s powernet. All APCs have lost power. Gravity systems likely to fail.", "Systems Power Failure", new_sound = 'sound/AI/attention.ogg')
+			for(var/obj/machinery/power/apc/A in GLOB.apcs)
+				var/area/current_area = get_area(A)
+				if((current_area.type in skipped_areas_apc) || !is_station_level(A.z))
+					continue
+				var/obj/item/stock_parts/cell/C = A.get_cell()
+				if(C)
+					C.charge = 0
+				current_area.power_change()
+			log_and_message_admins("Power has been drained from all APCs.")
+		if("Short out APCS")
+			var/list/skipped_areas_apc = list(
+			/area/engine/engineering,
+			/area/turret_protected/ai)
+			if(announce)
+				GLOB.event_announcement.Announce("Overload detected in [station_name()]'s powernet. Engineering, please repair shorted APCs.", "Systems Power Failure", new_sound = 'sound/AI/attention.ogg')
+			// break APC_BREAK_PROBABILITY% of all of the APCs on the station
+			var/affected_apc_count = 0
+			for(var/obj/machinery/power/apc/A in GLOB.apcs)
+				// skip any APCs that are too critical to disable
+				var/area/current_area = get_area(A)
+				if((current_area.type in skipped_areas_apc) || !is_station_level(A.z))
+					continue
+				// if we are going to break this one
+				if(prob(APC_BREAK_PROBABILITY))
+					// if it has internal wires, cut the power wires
+					if(A.wires)
+						if(!A.wires.is_cut(WIRE_MAIN_POWER1))
+							A.wires.cut(WIRE_MAIN_POWER1)
+						if(!A.wires.is_cut(WIRE_MAIN_POWER2))
+							A.wires.cut(WIRE_MAIN_POWER2)
+					// if it was operating, toggle off the breaker
+					if(A.operating)
+						A.toggle_breaker()
+					// no matter what, ensure the area knows something happened to the power
+					current_area.power_change()
+					affected_apc_count++
+			log_and_message_admins("APC Short event shorted out [affected_apc_count] APCs.")
 
 /proc/power_restore(announce=TRUE)
-	if(announce)
-		GLOB.event_announcement.Announce("Power has been restored to [station_name()]. We apologize for the inconvenience.", "Power Systems Nominal", new_sound = 'sound/AI/poweron.ogg')
-
-	// recharge the APCs
-	for(var/thing in GLOB.apcs)
-		var/obj/machinery/power/apc/A = thing
-		if(!is_station_level(A.z))
-			continue
-		var/obj/item/stock_parts/cell/C = A.get_cell()
-		if(C)
-			C.give(C.maxcharge)
+	switch(alert("What Would You Like to Do?", "Make All Areas Powered", "Power all APCs", "Repair all APCs", "Repair and Power APCs"))
+		if("Power all APCs")	//Power without Repairing
+			if(announce)
+				GLOB.event_announcement.Announce("All operational APCs on [station_name()] have been fully charged.", "Power Systems Nominal", new_sound = 'sound/AI/poweron.ogg')
+			var/affected_apc_count = 0
+			for(var/obj/machinery/power/apc/A in GLOB.apcs)
+				var/obj/item/stock_parts/cell/C = A.get_cell()
```

Comment:
I would avoid getting this cell until it is needed on line 87-88. There is a possibility that you get this cell, and then `continue` because the APC is not on the station z-level, thus doing a pointless operation.